src/resources/dart-3-migration.md

```diff
@@ -0,0 +1,512 @@
+---
+
+Dart 3 is a new major release of Dart, that brings new core capabilities like
+records, patterns, and classes. Alongside these new capabilities the release
+also contains a number of changes, that may break existing code. The present
+document is intended to be a migration guide that may be helpful in resolving
+these.
+
+## Introduction
+
+### Unversioned vs versioned changes
+
+The potentially breaking changes listed below fall into one of two categories:
+
+* **Unversioned changes**: These are changes which are affect any Dart code,
+which is used after upgrading to a Dart 3 SDK. There is no way to "turn off"
+these changes.
+
+* **Versioned changes**: Versioned changes are different, 
+and only apply when the package/app has selected the 
+Dart 3.0 (or above) language version. 
+The language version [is derived](/guides/language/evolution#language-version-numbers)
+from the `sdk` lower-constraint in the `pubspec.yaml` file. An SDK constraint like
+this does *not* apply the Dart 3 versioned changes:
+    ```yaml
+    environment:
+      sdk: '>=2.14.0 <3.0.0'
+    ```
+    But an SDK constraint like this does:
+    ```yaml
+    environment:
+      sdk: '>=3.0.0 <4.0.0'
+    ```
+
+Note that to use the new Dart 3 features you have to update the language version
+to 3.0, and thus you'll also get the Dart 3 versioned changes at the same time.
+
+### Dart 3 backwards compatibility
+
+Many packages and apps migrated to use null safety with Dart 2.12 or later are
+likely be backwards compatible with Dart 3. This is possible as for any package where
+the lower bound of the SDK constraint is 2.12.0 or higher, pub will allow
+resolution even when the upper bound is limited to versions below 3.0.0. For
+example, a package with the following constraint will be allowed to resolve with
+a Dart 3.x SDK, as pub will re-interpret the upper-constraint `<3.0.0` as `<4.0.0`
+when the lower constraint is `2.12` or higher:
+
+```yaml
+environment:
+  sdk: '>=2.14.0 <3.0.0'           # This is interpreted as '>=2.14.0 <4.0.0'
+```
+
+This allows developers to use Dart 3 sound null safety with packages that have
+been migrated to 2.12 null safety without needing a second migration, unless
+the code is affected by any other Dart 3 changes.
+
+### Testing for impact
+
+To understand if your source code is impacted by any Dart 3 changes, 
+use these steps:
+
+```terminal
+$ dart --version    # make sure this reports 3.0.0 or higher
+$ dart pub get      # this should resolve without issues
+$ dart analyze      # this should pass without errors
+```
+
+If the `pub get` step fails, try to upgrade your dependencies
+to see if more recent versions might support Dart 3:
+
+```terminal
+$ dart pub upgrade
+$ dart analyze      # this should pass without errors
+```
+
+Or if needed also include major versions upgrades:
+```terminal
+$ dart pub upgrade --major-versions
+$ dart analyze      # this should pass without errors
```

Comment:
```suggestion
Or, if needed, also include [major versions](tools/pub/cmd/pub-upgrade#--major-versions) upgrades:
```terminal
$ dart pub upgrade --major-versions
$ dart analyze      # This should pass without errors.
```